mixer/pkg/runtime2/routing/table.go

```diff
@@ -0,0 +1,170 @@
+	InputGroups []*InputGroup
+
+	// Maximum number of instances that can be created from this entry.
+	maxInstances int
+
+	// FriendlyName is the friendly name of this configured handler entry. Used for monitoring/logging purposes.
+	FriendlyName string
+
+	// Perf counters for keeping track of dispatches to adapters/handlers.
+	Counters DestinationCounters
+}
+
+// InputGroup is a set of instances that needs to be sent to a handler, grouped by a condition expression.
+type InputGroup struct {
+	// ID of the InputGroup. IDs are reused every time a table is recreated. Used for debugging.
+	ID uint32
+
+	// Condition for applying this input group.
+	Condition compiled.Expression
+
+	// TODO(Issue #2139): This should be removed when we stop doing resource-type based checks.
+	// ResourceType is the resource type condition for this input group.
+	ResourceType config.ResourceType
+
+	// Builders for the instances in this input set for each instance that should be applied.
+	Builders []template.InstanceBuilderFn
+
+	// Mappers for attribute-generating adapters that map output attributes into the main attribute set.
+	Mappers []template.OutputMapperFn
+}
+
+var emptyTable = &Table{ID: -1}
+
+// Empty returns an empty routing table.
+func Empty() *Table {
+	return emptyTable
+}
+
+// GetDestinations returns the set of destinations (handlers) for the given template variety and for the given namespace.
+func (t *Table) GetDestinations(variety tpb.TemplateVariety, namespace string) *NamespaceTable {
+	destinations, ok := t.entries[variety]
+	if !ok {
+		log.Debugf("No destinations found for variety: table='%d', variety='%d'", t.ID, variety)
+
+		return emptyDestinations
+	}
+
+	destinationSet := destinations.entries[namespace]
+	if destinationSet == nil {
+		log.Debugf("no rules for namespace, using defaults: table='%d', variety='%d', ns='%s'", t.ID, variety, namespace)
+		destinationSet = destinations.defaultSet
+	}
+
+	return destinationSet
+}
+
+// Count returns the number of entries contained.
+func (d *NamespaceTable) Count() int {
+	return len(d.entries)
+}
+
+// Entries in the table.
+func (d *NamespaceTable) Entries() []*Destination {
+	return d.entries
+}
+
+// MaxInstances returns the maximum number of instances that can be built from this Destination.
+func (d *Destination) MaxInstances() int {
+	return d.maxInstances
+}
+
+// used during building to recalculate maxInstances, after a modification.
+func (d *Destination) recalculateMaxInstances() {
+	c := 0
+	for _, input := range d.InputGroups {
+		c += len(input.Builders)
+	}
+
+	d.maxInstances = c
+}
+
+// Matches returns true, if the instances from this input set should be used for the given attribute bag.
+func (i *InputGroup) Matches(bag attribute.Bag) bool {
+	if i.Condition == nil {
+		return true
+	}
+
+	matches, err := i.Condition.EvaluateBoolean(bag)
+	if err != nil {
+		log.Warnf("input set condition evaluation error: id='%d', error='%v'", i.ID, err)
```

Comment:
I'm a bit worried about this error path & log message. This doesn't seem like something the operator can control, right?

Reply:
Discussed this offline. Essentially, this can happen under limited circumstances (e.g. the attributes referenced in the expressions are not found). So, there is operator control over these.

We can do a better job with our log messages, to help the operator. But that is probably best done holistically by ensuring we have a good traceability story (i.e. similar to the C# compiler error codes perhaps? Example: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0234)

I've opened this to track it: https://github.com/istio/istio/issues/3027